### PR TITLE
Nested Cosmos Db compatibility

### PR DESCRIPTION
**Pull request by countincognito**
There seems to be an issue when using nested queries (i.e. >1 level deep), such as Sum() and Any(), which causes the query provider in the Azure Cosmos SDK to generate invalid SQL. I was able to fix it by overriding empty parameter expression names. Apologies if I have missed anything.

### CODE REVIEW

File: src/System.Linq.Dynamic.Core/ParameterExpressionHelper.cs [hunk 1]
```diff
@@ -4,9 +4,38 @@ namespace System.Linq.Dynamic.Core
 {
     internal static class ParameterExpressionHelper
     {
-        public static ParameterExpression CreateParameterExpression(Type type, string name)
+        internal const int RandomWordLength = 6;
+
+        public static ParameterExpression CreateParameterExpression(Type type, string name, bool renameEmpty = false)
+        {
+            string paramName = name;
+            if (renameEmpty && IsNullOrWhiteSpace(paramName))
```
[the remark this answers is not in the document]
Comment (countincognito) [reply src/System.Linq.Dynamic.Core/ParameterExpressionHelper.cs]: I tried that originally, but unfortunately it doesn't exist for the .NET 3.5 version. How would you prefer to proceed?

File: src-console/ConsoleApp_netcore2.1_CosmosDb/ConsoleApp_netcore2.1_CosmosDb.csproj [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
+    <RootNamespace>ConsoleAppCosmosDb</RootNamespace>
+    <AssemblyName>ConsoleAppCosmosDb</AssemblyName>
+  </PropertyGroup>
+    
+    <ItemGroup>
+        <PackageReference Include="Microsoft.Azure.DocumentDB.Core" Version="2.9.4" />
```
[the remark this answers is not in the document]
Comment (countincognito) [reply src-console/ConsoleApp_netcore2.1_CosmosDb/ConsoleApp_netcore2.1_CosmosDb.csproj]: Done. Added one for .NET 4.5.2.

File: src/System.Linq.Dynamic.Core/ParsingConfig.cs [hunk 3]
```diff
@@ -134,6 +142,13 @@ public IQueryableAnalyzer QueryableAnalyzer
         /// </summary>
         public bool RenameParameterExpression { get; set; } = false;
 
+        /// <summary>
+        /// Prevents any System.Linq.Expressions.ParameterExpression.Name from being empty.
```
[the remark this answers is not in the document]
Comment (countincognito) [reply src/System.Linq.Dynamic.Core/ParsingConfig.cs]: Done.

File: src/System.Linq.Dynamic.Core/ParameterExpressionHelper.cs [hunk 4]
```diff
@@ -4,9 +4,38 @@ namespace System.Linq.Dynamic.Core
...
+            {
+                if (!char.IsWhiteSpace(value[i]))
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        internal static string GenerateRandomWord()
```
[the remark this answers is not in the document]
Comment (countincognito) [reply src/System.Linq.Dynamic.Core/ParameterExpressionHelper.cs]: Now generating words of 16 characters (just to be sure).

**Comment by countincognito** [comment]
> 
> Thank you for this PR, I never thought this library could be used for CosmosDB.
> 

My pleasure - such a great library, I only hope my contribution is up to standard. It's been phenomenally useful for our project with Cosmos Db (we use it in conjunction with the Cosmonaut library). This issue with the nested queries was the final hurdle for us to allow for more complicated queries.

File: src/System.Linq.Dynamic.Core/ParameterExpressionHelper.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (StefH) [reply src/System.Linq.Dynamic.Core/ParameterExpressionHelper.cs]: Keep your code (maybe add a comment).

[the remark this answers is not in the document]
Comment (countincognito) [reply src/System.Linq.Dynamic.Core/ParameterExpressionHelper.cs]: Done.

**Comment by countincognito** [comment]
@StefH Any further changes required?

**Comment by StefH** [comment]
@countincognito Thanks for this PR and your reworked code.

I'm doubting if I should just add this to the next version 1.0.21 / 1.0.22 or that I should use a higher version like 1.1.0

**Comment by countincognito** [comment]
> @countincognito Thanks for this PR and your reworked code.
> 
> I'm doubting if I should just add this to the next version 1.0.21 / 1.0.22 or that I should use a higher version like 1.1.0

Thanks for approving it. Glad the changes will be of use.

**Pull request merged**